jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/PacketHandler.kt

```diff
@@ -0,0 +1,201 @@
+        return try {
+            adaptiveSourceProjection.rewriteRtp(packetInfo)
+
+            // The rewriteRtp operation must not modify the VP8 payload.
+            if (ENABLE_PAYLOAD_VERIFICATION) {
+                val expected = packetInfo.payloadVerification
+                val actual = videoPacket.payloadVerification
+                if ("" != expected && expected != actual) {
+                    logger.warn("Payload unexpectedly modified! Expected: $expected, actual: $actual")
+                }
+            }
+            true
+        } catch (e: RewriteException) {
+            logger.warn("Failed to rewrite a packet.", e)
+            false
+        }
+    }
+
+    /**
+     * @return true if a packet should be accepted according to the current configuration.
+     */
+    fun accept(packetInfo: PacketInfo): Boolean {
+        val videoPacket = packetInfo.packetAs<VideoRtpPacket>()
+        val adaptiveSourceProjection = adaptiveSourceProjectionMap[videoPacket.ssrc]
+        if (adaptiveSourceProjection == null) {
+            logger.debug { "Dropping an RTP packet for an unknown SSRC: ${videoPacket.ssrc}" }
+            numDroppedPacketsUnknownSsrc.incrementAndGet()
+            return false
+        }
+        return adaptiveSourceProjection.accept(packetInfo)
+    }
+
+    /**
+     * @return true if [rtcpSrPacket] should be accepted, false otherwise.
+     *
+     * Filters out packets that match one of the streams that this instance manages, but don't match the target SSRC.
+     * Allows packets for streams not managed by this instance.
+     */
+    fun accept(rtcpSrPacket: RtcpSrPacket): Boolean {
+        val adaptiveSourceProjection = adaptiveSourceProjectionMap[rtcpSrPacket.senderSsrc]
+
+        // This is probably for an audio stream. In any case, if it's for a stream which we are not forwarding it
+        // will be stripped off at a later stage (in RtcpSrUpdater).
+        return adaptiveSourceProjection == null ||
+            // We only accept SRs for the SSRC that we're forwarding with.
+            adaptiveSourceProjection.targetSsrc == rtcpSrPacket.senderSsrc
+    }
+
+    /**
+     * @return true if the packet was transformed successfully, false otherwise.
+     */
+    fun transformRtcp(rtcpSrPacket: RtcpSrPacket): Boolean =
+        adaptiveSourceProjectionMap[rtcpSrPacket.senderSsrc]?.rewriteRtcp(rtcpSrPacket) ?: false
+
+    /**
+     * Utility method that looks-up or creates the adaptive source projection of
+     * a source.
+     */
+    private fun lookupOrCreateAdaptiveSourceProjection(singleAllocation: SingleAllocation): AdaptiveSourceProjection? {
+        val source = singleAllocation.mediaSource
+        val endpointID = singleAllocation.endpointId
+        if (source == null) {
+            return null
+        }
+        synchronized(adaptiveSourceProjectionMap) {
+            var adaptiveSourceProjection = adaptiveSourceProjectionMap[source.primarySSRC]
```

Comment:
Using `adaptiveSourceProjection` as a `var` both for the result of the lookup and for the newly-constructed object doesn't seem very Kotlin-y, but I'm not sure what a better idiom would be.  Maybe have the result of the lookup be `val existingProjection` or something?

Reply:
Cleaned up